core/src/main/java/software/amazon/awssdk/interceptor/ExecutionInterceptorChain.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.interceptor;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import software.amazon.awssdk.SdkRequest;
+import software.amazon.awssdk.SdkResponse;
+import software.amazon.awssdk.annotation.SdkInternalApi;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.interceptor.context.DefaultFailedExecutionInterceptorContext;
+import software.amazon.awssdk.interceptor.context.InterceptorContext;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A wrapper for a list of {@link ExecutionInterceptor}s that ensures the interceptors are executed in the correct order as it
+ * is documented in the {@link ExecutionInterceptor} documentation.
+ *
+ * Interceptors are invoked in forward order up to {@link #beforeTransmission} and in reverse order after (and including)
+ * {@link #afterTransmission}. This ensures the last interceptors to modify the request are the first interceptors to see the
+ * response.
+ */
+@SdkInternalApi
+public class ExecutionInterceptorChain {
```

Comment:
As mentioned above: 

I did it because the parameters to the ExecutionInterceptor methods are interfaces, so there's no easy way to modify them as part of the modify* methods. It's not impossible, but this just seemed simpler. I'm open to alternative ideas.